Added missing 'time' for a field manager that server-side-applied same configuration

waltforme opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind bug

#### What this PR does / why we need it:
The 'time' stanza in `metadata.managedFields` is missing under some circumstances after a server-side apply. This PR tries to get the missing 'time' stanza back. More details are documented in #127938.

#### Which issue(s) this PR fixes:
Fixes #127938

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?

```release-note
Fixed an issue in `metadata.managedFields`. Before this fix, the `time` stanza for a 2nd field manager was not set if that 2nd field manager server-side applies the same configuration as the 1st manager.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

waltforme committed:
Added missing 'time' for a field manager who applied same configuration

Signed-off-by: Jun Duan <jun.duan.phd@outlook.com>

Jefftree commented:
/assign @seans3 
/cc @jpbetz 
/triage accepted

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater.go:
```diff
@@ -72,9 +72,16 @@ func (f *managedFieldsUpdater) Apply(liveObj, appliedObj runtime.Object, managed
 	if object != nil {
 		managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
 	} else {
+		if managerInFields && !managerInTimes {
```
I'm not very familiar with this code, but it looks strange to put this logic in the branch that is calling `RemoveObjectManagedFields` ... are we sure this is the branch where a new manager is *adding* co-ownership of some fields?

this definitely needs really good unit tests around the scenario being exercised

waltforme replied:
Thanks for the prompt review!

I have the same feeling that the two branches are a little confusing to follow. Let me try my best to clarify.
- Since `managedFieldsUpdater` implements the `Manager` interface, its method `(f *managedFieldsUpdater) Apply` should return ‘the new object with managedFields removed, and the object's new proposed managedFields **separately**’:
   https://github.com/kubernetes/kubernetes/blob/7ee17ce9b7c2a22e63e2bbd79d48d3fe349a9386/staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/manager.go#L45-L51
- The `else` branch says: Since the live object is not touched (signaled by `object == nil`, detailed below), the merged object should be the same as the live object. So first make a deep copy of `liveObj`, then remove its managedFields to fit the `Manager` interface.

Details on `object == nil`:
These lines say that `nil` signals the situation that a merged object and the current live object being the same.
https://github.com/kubernetes/kubernetes/blob/7ee17ce9b7c2a22e63e2bbd79d48d3fe349a9386/vendor/sigs.k8s.io/structured-merge-diff/v4/merge/update.go#L209-L211
In our use case, the 2nd field manager applies the same configuration as the 1st, so the merged object is indeed the same as the current live object, so a `nil` is indeed returned to signal that.

BTW I believe there was a typo in the go doc comments of the `Manager` interface’s `Apply` method. Latest push fixed the typo.

Fully agree that we need unit tests here. Will work on that.

liggitt replied:
I think the logic in this PR is correct, but needs commenting to explain.

I'd suggest:
```go
	if object != nil {
		// non-nil object means the apply operation modified the object, so update the manager's timestamp
		managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
	} else {
		// nil object means the apply operation did not modify the input object
		// clone the input object and return it without managed fields
		object = liveObj.DeepCopyObject()
		RemoveObjectManagedFields(object)

		if _, managerInFields := managed.Fields()[fieldManager]; managerInFields {
			if _, managerInTimes := managed.Times()[fieldManager]; !managerInTimes {
				// if the manager owns fields, ensure it has an associated time.
				// a no-op apply can add co-ownership of existing field values, so record the time that occurred.
				managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
			}
		}
}
```

waltforme committed:
Corrected go doc comments for Apply method of Manager

Signed-off-by: Jun Duan <jun.duan.phd@outlook.com>

sftim commented:
Would this PR need a changelog entry?

MikeSpreitzer commented:
@sftim: how much would be appropriate to put in the change log entry? Just "Fixed Issue 127938", or a description of the behavior change?

sftim commented:
Try something like:

```markdown
Fixed an issue in the frobnicator. Before this fix, the `time` field was not set correctly if your
time machine was travelling below 141 kilometers per hour.
```

waltforme commented:
> Try something like:
> 
> ```
> Fixed an issue in the frobnicator. Before this fix, the `time` field was not set correctly if your
> time machine was travelling below 141 kilometers per hour.
> ```

@sftim Added. Thanks for the nice example!

waltforme committed:
Update unit tests for managedfieldsupdater

Signed-off-by: Jun Duan <jun.duan.phd@outlook.com>

waltforme commented:
As suggested by @liggitt, the latest push added unit test. The added test `TestCoOwningManagedFieldsByApplyingSameObjResultsInNewManagerTime` checks:
- The 2nd field manager's server-side apply doesn't change the 1st field manager's managed fields;
- The 2nd field manager co-own the managed fields together with the 1st field manager;
- The 2nd field manager has its own 'time' stanza.

Latest push also made a few small corrections to the previously existing unit tests.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater_test.go:
```diff
@@ -335,7 +335,7 @@ func TestTakingOverManagedFieldsDuringUpdateDoesNotModifyPreviousManagerTime(t *
 		"data": {
 			"key_a": "value",
-			"key_b": value"
+			"key_b": "value"
```
huh... these test typos are distressing... did you just notice these and fix them while you were here, or did fixing these reveal that these tests were broken?

liggitt replied:
non-blocking for this PR, but cc @jpbetz for visibility to how accepting yaml parsing is of malformed json input sent as yaml. This was parsing as:

```
{
  "apiVersion": "v1",
  "data": {
    "key_a": "value",
    "key_b": "value\""
  },
  "kind": "ConfigMap",
  "metadata": {
    "name": "configmap"
  }
}
```

waltforme replied:
The former, I just noticed these so fixed them. I was wondering why these typos didn't break the testing, and now I get the answer from your explanations above. Thanks!

waltforme commented:
The force-push to https://github.com/kubernetes/kubernetes/commit/ce32cdca21d7a63d5aff5426e2bcf96ee383fbf6 included the suggested changes from @liggitt, also squashed those changes to appropriate commit.

I messed up one unrelated commit in the previous force-push to https://github.com/kubernetes/kubernetes/commit/6d5c8023158540801d53f2fa115edc15204b95d1, and the bot added `area/kubelet` and `sig/node` labels to this PR. I corrected git history afterwards, but I could not remove the two labels by myself. I'm so sorry if this introduces any confusion. **Just want to clarify: This PR is not related to kubelet or sig-node.**

MikeSpreitzer commented:
/test pull-kubernetes-e2e-gce-100-performance
(to get a baseline for comparison with #128087 and #128974)

MikeSpreitzer commented:
/remove-area kubelet

MikeSpreitzer commented:
/remove-sig node

waltforme commented:
Thanks @MikeSpreitzer !

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater.go:
```diff
@@ -72,11 +72,23 @@ func (f *managedFieldsUpdater) Apply(liveObj, appliedObj runtime.Object, managed
 		object = liveObj.DeepCopyObject()
 		RemoveObjectManagedFields(object)
+
+		if _, managerInFields := managed.Fields()[fieldManager]; managerInFields {
```
@jpbetz how does partitioning of field manager among subresources work? does managed.Fields() only return entries relevant to the subresource for this request?

jpbetz replied:
The field manager relies on the `GetResetFields()` ([example](https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/pkg/registry/core/pod/strategy.go#L202)) of the subresource to indicate which fields the subresource applies to. 

'GetResetFieldsFilter()' was also added a in 1.32 to make it easier to declare fields for subresources ([example](https://github.com/kubernetes/kubernetes/blob/810e9e212ec5372d16b655f57b9231d8654a2179/pkg/registry/core/pod/strategy.go#L339C28-L343))

liggitt replied:
The specific test I'm thinking of is:

1. manager `foo` writes to pod, gets a managedFields entry for itself
2. manager `bar` writes to pods/status, gets a managedFields entry for itself under the status subresource
3. manager `bar` does no-op write to pod (e.g. empty patch `{}`) so that it doesn't actually own any fields

Will the change in this PR make processing of step 3 see and get confused by the managed fields entry from 2?

liggitt commented on staging/src/k8s.io/apimachinery/pkg/util/managedfields/internal/managedfieldsupdater.go:
```diff
@@ -72,11 +72,23 @@ func (f *managedFieldsUpdater) Apply(liveObj, appliedObj runtime.Object, managed
+			if _, managerInTimes := managed.Times()[fieldManager]; !managerInTimes {
+				// if the manager owns fields, ensure it has an associated time.
+				// a no-op apply can add co-ownership of existing field values, so record the time that occurred.
+				managed.Times()[fieldManager] = &metav1.Time{Time: time.Now().UTC()}
```
@jpbetz for existing managed entries that got created without timestamps, once this bug fix releases, the *next* no-op update will set the time to `now` on a pre-existing managed field entry. That's a little weird, and probably ok, but I wanted to call it out explicitly as a side-effect of the fix.

jpbetz replied:
We fixed a few spurious resourceVersion bumps issues this year:

- https://github.com/kubernetes/kubernetes/pull/125317
- https://github.com/kubernetes/kubernetes/pull/125263

So I'm keeping an eye out of anything that changes an apply request *before* the object is compared with the stored resource to decide if a write is needed. But this happens *after*  `fieldmanager.Apply()` so appears safe, at least from that perspective.

MohammadAlavi1986 commented:
I believe there is a case that is not covered in this PR. If a manager performs a server-side apply (SSA) with a configuration that does not change the object, and then later performs another SSA with a different configuration that also does not change the object (but changes the list of field ownership for this manager), the `time` field is not updated for the second SSA.

Here's an example:

```
k apply --server-side --field-manager m1 --show-managed-fields -o yaml  -f - <<'EOF'
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: default
data:
  key1: value1
  key2: value2
EOF

# a new entry is added for m2 field manager with the correct time
k apply --server-side --field-manager m2 --show-managed-fields -o yaml  -f - <<'EOF'
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: default
data:
  key1: value1
EOF

# list of fields owned by m2 manager has changed, but time is NOT updated for m2 field manager
k apply --server-side --field-manager m2 --show-managed-fields -o yaml  -f - <<'EOF'
apiVersion: v1
kind: ConfigMap
metadata:
  name: test
  namespace: default
data:
  key1: value1
  key2: value2
EOF
```
